### ISSUE TITLE
RigidBody2D not colliding properly with CollisionShape2D (set to one_way_collision: on)

### ISSUE THREAD

**Issue by hiulit** [godotengine/godot#27593]
**Godot version:**

Godot Engine 3.1.stable.official

**OS/device including version:**

macOs Mojave 10.14.3
Intel Iris Pro 1536 MB

**Issue description:**

I'm using GLES2.

Scene tree:

```
Node2D
├── TileMap
├── KinematicBody2D
├── RigidBodyBody2D
└── Node2D
    ├── StaticBody2D
    ├── StaticBody2D
    ├── etc
```

The `TileMap` has a `tileset.tres` **Tile Set** with a **Single Tile** with `selected_collision_one_way` set to `on`. If I turned if `off` it seems to work.

I've also tried using a bunch of `StaticBody2D` with `CollisionShape2D` and I get the same behaviour. When `one_way_collision` is `on` the `RigidBodyBody2D`  doesn't collide properly. But when it's `off` it seems to work fine.

The `RigidBodyBody2D` is set to `Mode: Rigid`. If I change it to `Mode: Character` seems to work fine with either `one_way_collision` on and off.

![rigidbody2d-collisions](https://user-images.githubusercontent.com/481475/55395212-665b3780-5541-11e9-8646-3e1aeca78a13.gif)

**Steps to reproduce:**

Download the zip and try to play with the `TileMap` and the `StaticBody2D` changing `one_way_collision` to `on/off`. Also try playing with the `RigidBodyBody2D` modes.

**Minimal reproduction project:**

[test-rigidbody2d-collisions-one-way.zip](https://github.com/godotengine/godot/files/3033408/test-rigidbody2d-collisions-one-way.zip)

**UPDATE**

Version with the body's center of mass fixed.

[test-rigidbody2d-collisions-one-way-fixed-body-center-mass.zip](https://github.com/godotengine/godot/files/3058057/test-rigidbody2d-collisions-one-way-fixed-body-center-mass.zip)

**Comment by Xrayez** [comment]
The rigid body's shape isn't aligned with body's center of mass, which may or may not be relevant to this issue, also see #12353.

**Comment by hiulit** [comment]
Thanks for pointing this out @Xrayez . I changed the body's center mass and now the interaction with `CollisionShape2D` is better, but it still goes through it when it's set to `one_way_collision`.

The `RigidBody2D` are set to `Rigid` mode.

![rigidbody2d-1](https://user-images.githubusercontent.com/481475/55486852-4ef97880-562d-11e9-9d2e-72594f90b49a.gif)

![rigidbody2d-2](https://user-images.githubusercontent.com/481475/55486860-54ef5980-562d-11e9-8f24-42a40bfff3bd.gif)

The left one goes through a `TileMap` that has a `tileset.tres` **Tile Set** with a **Single Tile** with `selected_collision_one_way` set to `on`.

The middle one stops on a `StaticBody2D` with a `CollisionShape2D` with `one_way_collision` set to `on`.

The right one stops on a `StaticBody2D` with a `CollisionShape2D` with `one_way_collision` set to `off`.

**Comment by hiulit** [comment]
Here's another example to show that they go through with wither `one_way_collision set` to `on`.

![rigidbody2d-3](https://user-images.githubusercontent.com/481475/55487283-1a39f100-562e-11e9-9d16-b5b5f4d7cebf.gif)

**Comment by hiulit** [comment]
Here's the test project with the fixed body's center of mass.

[test-rigidbody2d-collisions-one-way-fixed-body-center-mass.zip](https://github.com/godotengine/godot/files/3058054/test-rigidbody2d-collisions-one-way-fixed-body-center-mass.zip)

**Comment by warner-programming** [comment]
I have the same issue:
http://www.giphy.com/gifs/cMb9ViHmeLogq7v9Ts

It looks as if as soon as one of the contact points gets below the one way platform top, the rigidbody falls through. I can imagine that a solution would be to check the previous state of the object instead of the current one: if it came from above, it should collide.

Here is the source:
http://tiny.cc/6dtmfz

**Comment by madmiraal** [comment]
Fixed with #39880 and cherry-picked for 3.2.3.
